### PR TITLE
SmartStore Plugin v1

### PR DESCRIPTION
**Pull request by Nisaba**

### CODE REVIEW

File: Models/ConfigurationModel.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        [Url]
+        //[Required]
+        public string ApiUrl { get; set; }
+
+        [LocalizedDisplay("*ApiKey")]
+        //[Required]
+        public string ApiKey { get; set; }
+
+        [LocalizedDisplay("*ApiSecret")]
+        //[Required]
```
Comment (SwissBitcoinPay) [inline Models/ConfigurationModel.cs:17]: ApiSecret should not be required.
If not defined, the way to secure the webhook is as follow:
1. In the webhook handler, make a GET request to `/checkout/${invoiceId}`
2. Check if the `isPaid` value is effectively `true`
3. Set the order as Paid

File: Configuration/SwissBitcoinPaySettings.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using Smartstore.Core.Configuration;
+
+namespace SmartStore.SwissBitcoinPay.Settings
+{
+    public class SwissBitcoinPaySettings : ISettings
+    {
+        /// <summary>
+        /// The url of the Swiss Bitcoin Pay API
+        /// </summary>
+        public string ApiUrl { get; set; }
```
Comment (SwissBitcoinPay) [inline Configuration/SwissBitcoinPaySettings.cs:10]: The parameter does not have to be configurable nor visible to users. Is it possible to remove it?

File: Models/SwissBitcoinPayInvoiceModel.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+namespace SmartStore.SwissBitcoinPay.Models
+{
+
+    public struct SwissBitcoinPayInvoiceModel
+    {
+        public string title;
+        public string description;
+        public decimal amount;
+        public string unit;
+        //public bool onChain;
```
Comment (SwissBitcoinPay) [inline Models/SwissBitcoinPayInvoiceModel.cs]: You should add a boolean setting "Accept Onchain payments"

File: Models/SwissBitcoinPayInvoiceModel.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        //public int delay;
```
Comment (SwissBitcoinPay) [inline Models/SwissBitcoinPayInvoiceModel.cs]: You can delete this comment, and set delay to `60` for all created invoices (60 minutes)

File: Services/SwissBitcoinPayService.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+				}
+
+				string sRep;
+				using (var rep = req.GetResponse())
+				{
+					using (var rdr = new StreamReader(rep.GetResponseStream()))
+					{
+						sRep = rdr.ReadToEnd();
+					}
+				}*/
```
Comment (SwissBitcoinPay) [inline Services/SwissBitcoinPayService.cs]: If not used, you can delete this comment

File: Services/SwissBitcoinPayService.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+        public string CreateInvoice(SwissBitcoinPaySettings settings, PaymentDataModel paymentData)
+        {
+
+            try
+            {
+                var invoice = new SwissBitcoinPayInvoiceModel()
+				{
+                    title = paymentData.Description,
+                    description = $"{paymentData.BuyerName}#Order:{paymentData.OrderID}#Store:{paymentData.StoreID.ToString()}",
```
Comment (SwissBitcoinPay) [inline Services/SwissBitcoinPayService.cs]: BuyerName seems to be empty, when creating an invoice.
You can also remove the StoreID

```suggestion
                    description = $"{paymentData.BuyerName} | Order {paymentData.OrderID}",

```

File: Providers/SwissBitcoinPayProvider.cs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                result.AuthorizationTransactionResult = apiService.CreateInvoice(settings, new PaymentDataModel()
+                {
+                    CurrencyCode = _currencyService.PrimaryCurrency.CurrencyCode,
+                    Amount = processPaymentRequest.OrderTotal,
+                    BuyerEmail = "" + myCustomer.Email,
+                    BuyerName = myCustomer.FirstName + " " + myCustomer.LastName,
+                    OrderID = processPaymentRequest.OrderGuid.ToString(),
+                    StoreID = myStore.Id,
+                    CustomerID = myCustomer.Id,
+                    Description = "From " + myStore.Name,
```
Comment (SwissBitcoinPay) [inline Providers/SwissBitcoinPayProvider.cs:96]: For easier reading for the shop owner in their dashboard, it's better to include a (short) order ID, like order number.

**Review by SwissBitcoinPay — CHANGES_REQUESTED** [review]
Some changes are required

File: Models/SwissBitcoinPayInvoiceModel.cs (hunk 4 above)
Comment (Nisaba) [reply Models/SwissBitcoinPayInvoiceModel.cs]: ok, done.

File: Services/SwissBitcoinPayService.cs (hunk 5 above)
Comment (Nisaba) [reply Services/SwissBitcoinPayService.cs]: ok, done.

File: Services/SwissBitcoinPayService.cs (hunk 6 above)
Comment (Nisaba) [reply Services/SwissBitcoinPayService.cs]: No, SmartStore Store ID is mandatory by SmartStore.
And the buyer name comes from SmartStore. If it's empty, check the customer settings in SmartStore

File: Models/SwissBitcoinPayInvoiceModel.cs (hunk 3 above)
Comment (Nisaba) [reply Models/SwissBitcoinPayInvoiceModel.cs]: How should this setting be managed in the plugin ?

File: Configuration/SwissBitcoinPaySettings.cs (hunk 2 above)
Comment (Nisaba) [reply Configuration/SwissBitcoinPaySettings.cs:10]: OK, done.
This parameter is hidden in the config page.

File: Providers/SwissBitcoinPayProvider.cs (hunk 7 above)
Comment (Nisaba) [reply Providers/SwissBitcoinPayProvider.cs:96]: The order ID in Smartstore is a GUID and is not short :)
The field in checkout payment page seems to shot to add something else

File: Models/ConfigurationModel.cs (hunk 1 above)
Comment (Nisaba) [reply Models/ConfigurationModel.cs:17]: Can we discuss about this ?

File: Models/SwissBitcoinPayInvoiceModel.cs (hunk 3 above)
Comment (Nisaba) [reply Models/SwissBitcoinPayInvoiceModel.cs]: OK, done.
I've added a flag in the merchant configuration to allow to accept onchain payments

File: Services/SwissBitcoinPayService.cs (hunk 6 above)
Comment (SwissBitcoinPay) [reply Services/SwissBitcoinPayService.cs]: Ok, at least you can use the | and put spaces around like the suggestion above, to make it more readable.

File: Services/SwissBitcoinPayService.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+                    unit = paymentData.CurrencyCode,
+					amount = paymentData.Amount,
+                    email = paymentData.BuyerEmail,
+                    emailLanguage = paymentData.Lang,
+                    redirectAfterPaid = paymentData.RedirectionURL,
+                    webhook = paymentData.WebHookURL,
+                    delay = 60,
+                    onChain = settings.AcceptOnChain
```
Comment (SwissBitcoinPay) [inline Services/SwissBitcoinPayService.cs]: Lastly I'd like to include a `customNote` so the merchant have a a distinguishable ID when seeing the list of his orders :

```suggestion
                    onChain = settings.AcceptOnChain,
                    extra = {
                      customNote: $"Order {paymentData.OrderID}"
                    }
```
(not sure about the synthax in C# to put a sub-value in an `extra` object)

image

**Review by SwissBitcoinPay — CHANGES_REQUESTED** [review]

File: Services/SwissBitcoinPayService.cs (hunk 8 above)
Comment (Nisaba) [reply Services/SwissBitcoinPayService.cs]: I don't see an extra field in the API documentation ?

Comment (SwissBitcoinPay) [reply Services/SwissBitcoinPayService.cs]: It's not documented yet, but it is implemented. You can use it.

File: Services/SwissBitcoinPayService.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+					amount = paymentData.Amount,
+                    email = paymentData.BuyerEmail,
+                    emailLanguage = paymentData.Lang,
+                    redirectAfterPaid = paymentData.RedirectionURL,
+                    webhook = paymentData.WebHookURL,
+                    delay = 60,
+                    onChain = settings.AcceptOnChain,
+                    extra = new SwissBitcoinPayExtraModel()
+                    {
+                        CustomNote = $"Order {paymentData.OrderID}"
```
Comment (SwissBitcoinPay) [inline Services/SwissBitcoinPayService.cs]: Doit être en camelCase :)

```suggestion
                        customNote = $"Order {paymentData.OrderID}"

```

**Comment by SwissBitcoinPay** [comment]
All good! Thanks

**Pull request merged**